src/qt/guiutil.h

```diff
@@ -273,6 +279,30 @@ namespace GUIUtil
      * Writes to debug.log short info about the used Qt and the host system.
      */
     void LogQtInfo();
+
+    /**
+     * Throwing an exception from a slot invoked by Qt's signal-slot connection
+     * mechanism is considered undefined behavior, unless it is handled within
+     * the slot. The shieldException() function should be used for exception
+     * handling within slots.
+     */
+    template <typename FunctionType> void shieldException(FunctionType&& f) {
+        bool ok = true;
+        try {
+            f();
+        } catch (const NonFatalCheckError& e) {
+            PrintExceptionContinue(&e, "Internal error");
+            ok = QMetaObject::invokeMethod(qApp, "handleNonFatalException", Qt::DirectConnection, Q_ARG(QString, QString::fromStdString(e.what())));
```

Comment:
Why can't it be `AutoConnection`? That way this could be used in a "worker" slot.

Otherwise I'd assert the current thread is the GUI thread at the function begin.

Reply:
I'd keep `Qt::DirectConnection` to be sure the exception is handled in the same thread where it was thrown.

As `QMessageBox` in `BitcoinApplication::handleNonFatalException()` has its own event loop, I think it is safe to call `shieldException()` from a non-GUI thread, no?

Reply:
>  has its own event loop, I think it is safe to call shieldException() from a non-GUI thread, no?

I don't think so, only the main (GUI) thread should deal with widgets.

> I'd keep `Qt::DirectConnection` to be sure the exception is handled in the same thread where it was thrown.

What you mean by handled? This is just catching the exception (which happens in the right thread) and then showing the message. If you want to block until the user presses OK then try `GUIUtil::blockingGUIThreadConnection.`

Reply:
> I don't think so, only the main (GUI) thread should deal with widgets.

You are right. [Fixed](https://github.com/bitcoin/bitcoin/pull/18897#issuecomment-627740852).